src/test/RecordingMetadata.tsx/RecordingLabelFields.test.tsx

```diff
@@ -0,0 +1,199 @@
+      mockValid = state;
+    },
+  };
+
+  it('renders correctly', async () => {
+    let tree;
+    await act(async () => {
+      tree = renderer.create(<RecordingLabelFields {...mockProps} />);
+    });
+    expect(tree.toJSON()).toMatchSnapshot();
+  });
+
+  it('displays all labels in form fields', () => {
+    render(<RecordingLabelFields {...mockProps} />);
+
+    const label1KeyInput = screen.getAllByDisplayValue('someLabel')[0];
+    const label1ValueInput = screen.getAllByDisplayValue('someValue')[0];
+    const label2KeyInput = screen.getAllByDisplayValue('anotherLabel')[0];
+    const label2ValueInput = screen.getAllByDisplayValue('anotherValue')[0];
+
+    expect(label1KeyInput).toHaveClass('pf-c-form-control');
+    expect(label1ValueInput).toHaveClass('pf-c-form-control');
+    expect(label2KeyInput).toHaveClass('pf-c-form-control');
+    expect(label2ValueInput).toHaveClass('pf-c-form-control');
+
+    expect(screen.getByText('Add Label')).toBeInTheDocument();
+  });
+
+  it('updates the label key when entering text into the Key input', () => {
+    render(<RecordingLabelFields {...mockProps} />);
+
+    const labelKeyInput = screen.getAllByLabelText('label key')[0];
+    userEvent.clear(labelKeyInput);
+
+    userEvent.type(labelKeyInput, 'someEditedKey');
+    expect(mockProps.labels[0].key).toBe('someEditedKey');
+  });
+
+  it('updates the label value when entering text into the Value input', () => {
+    render(<RecordingLabelFields {...mockProps} />);
+
+    const labelValueInput = screen.getAllByLabelText('label value')[0];
+
+    userEvent.clear(labelValueInput);
+
+    userEvent.type(labelValueInput, 'someEditedValue');
+
+    expect(mockProps.labels[0].value).toBe('someEditedValue');
+  });
+
+  it('adds a label when Add Label is clicked', () => {
+    render(<RecordingLabelFields {...mockProps} />);
+
+    expect(screen.getAllByLabelText('label key').length).toBe(2);
+    expect(screen.getAllByLabelText('label value').length).toBe(2);
+
+    userEvent.click(screen.getByText('Add Label'));
+
+    expect(screen.getAllByLabelText('label key').length).toBe(3);
+    expect(screen.getAllByLabelText('label value').length).toBe(3);
+  });
+
+  it('removes the correct label when Delete button is clicked', () => {
+    render(<RecordingLabelFields {...mockProps} />);
+
+    expect(screen.getAllByLabelText('label key').length).toBe(2);
+    expect(screen.getAllByLabelText('label value').length).toBe(2);
+
+    userEvent.click(screen.getAllByTestId('remove-label-button')[0]);
+
+    expect(screen.getAllByLabelText('label key').length).toBe(1);
+    expect(screen.getAllByLabelText('label value').length).toBe(1);
+
+    expect(screen.getAllByDisplayValue('someLabel').length).toBe(0);
+    expect(screen.getAllByDisplayValue('someValue').length).toBe(0);
+    expect(screen.getAllByDisplayValue('anotherLabel')[0]).toBeInTheDocument();
+    expect(screen.getAllByDisplayValue('anotherValue')[0]).toBeInTheDocument();
+  });
```

Comment:
These two tests are failing because the `Add` and `Delete` label buttons don't change the size of the `mockLabels` array.  `await waitFor(() => expect(screen.getAllByLabelText('label key').length).toBe(1));` and `await waitFor(() => expect(screen.getAllByLabelText('label key').length).toBe(1), { timeout: 10});` also results in the same output. The buttons work fine when manually testing them.

@hareetd Do you know if there's anything I should be doing differently when passing props or setting the `mockLabels` array with Jest?

```
 ● <RecordingLabelFields /> › adds a label when Add Label is clicked

    expect(received).toBe(expected) // Object.is equality

    Expected: 3
    Received: 2

      131 |     userEvent.click(screen.getByText('Add Label'));
      132 |
    > 133 |     expect(screen.getAllByLabelText('label key').length).toBe(3);
          |                                                          ^
      134 |     expect(screen.getAllByLabelText('label value').length).toBe(3);
      135 |   });
      136 |

      at Object.<anonymous> (src/test/RecordingMetadata.tsx/RecordingLabelFields.test.tsx:133:58)
      at TestScheduler.scheduleTests (node_modules/@jest/core/build/TestScheduler.js:333:13)
      at runJest (node_modules/@jest/core/build/runJest.js:387:19)
      at _run10000 (node_modules/@jest/core/build/cli/index.js:408:7)
      at runCLI (node_modules/@jest/core/build/cli/index.js:261:3)

  ● <RecordingLabelFields /> › removes the correct label when Delete button is clicked

    expect(received).toBe(expected) // Object.is equality

    Expected: 1
    Received: 2

      143 |     userEvent.click(screen.getAllByTestId('remove-label-button')[0]);
      144 |
    > 145 |     expect(screen.getAllByLabelText('label key').length).toBe(1);
          |                                                          ^
      146 |     expect(screen.getAllByLabelText('label value').length).toBe(1);
      147 |
      148 |     expect(screen.getAllByDisplayValue('someLabel').length).toBe(0);
```

Reply:
So the thing with RTL is that it's designed purely from a user perspective. In this situation where the parent component's state and state handlers are passed in as props to the component under test, all we need to confirm is that:

1) Is the passed-in prop data rendered correctly?
2) Do changes in the component result in the appropriate prop handler being called?

From your other tests, we can confirm the passed-in labels are indeed correctly rendered so all that's left to confirm is `props.setLabels` is called when the add and delete buttons are clicked. We assume that the parent component correctly updates the labels when `props.setLabels` is called and since we know `RecordingLabelFields` will render the updated labels correctly, the test for this workflow is complete. The actual behavior of `props.setLabels` can be tested in the unit tests for the parent components themselves. 

If you think about it from a user perspective, when interacting with the `RecordingLabelFields` component they see the rendered labels and then either press "Add Label" or the "X" to delete a label, all of which you've fully tested for. The user has no idea about the internal state or props of a component, they simply interact with the render as described in the previous sentence. 

This is one of the reasons why unit tests can be tricky with RTL. If Enzyme was still an option we would be able to easily mock the internal props and state of a component but RTL doesn't allow that, which is why RTL exists on a sort of continuum between unit testing and integration testing. You could test `RecordingLabelFields` and its parent components in the same test, where the "unit" in the unit test is no longer each isolated component but instead, the entire label workflow, which would help solve these state mocking issues. However, I still think it's worthwhile to isolate components and unit test them in a traditional manner.

Reply:
> In this situation where the parent component's state and state handlers are passed in as props to the component under test

Ah thanks, that makes sense. I've moved the part of the test that checks if the state is actually updated to the parent component tests, and changed the child component test to verify that `props.setLabels` is called with the right arguments.